tools/istio-docker.mk

```diff
@@ -91,42 +87,42 @@ docker.sidecar_injector: $(ISTIO_DOCKER)/sidecar-injector
 docker.proxy: tools/deb/envoy_bootstrap_tmpl.json
 docker.proxy: ${ISTIO_ENVOY_RELEASE_PATH}
 docker.proxy: $(ISTIO_OUT)/pilot-agent ${PROXY_JSON_FILES}
-docker.proxy: pilot/docker/Dockerfile.proxy pilot/docker/Dockerfile.proxy_debug
-	mkdir -p $(ISTIO_DOCKER)/proxy
-	cp $^ $(ISTIO_DOCKER)/proxy/
-ifeq ($(DEBUG_IMAGE),1)
-	cp ${ISTIO_ENVOY_DEBUG_PATH} $(ISTIO_DOCKER)/proxyd/envoy
-	time (cd $(ISTIO_DOCKER)/proxy && \
-		docker build -t $(HUB)/proxy:$(TAG) -f Dockerfile.proxy_debug .)
-else
-	cp ${ISTIO_ENVOY_RELEASE_PATH} $(ISTIO_DOCKER)/proxy/envoy
-	time (cd $(ISTIO_DOCKER)/proxy && \
+docker.proxy: pilot/docker/Dockerfile.proxy
+	mkdir -p $(DOCKER_BUILD_TOP)/proxy
+	# Not using $^ to avoid 2 copies of envoy
+	cp tools/deb/envoy_bootstrap_tmpl.json $(ISTIO_OUT)/pilot-agent ${PROXY_JSON_FILES} pilot/docker/Dockerfile.proxy $(DOCKER_BUILD_TOP)/proxy/
+	cp ${ISTIO_ENVOY_RELEASE_PATH} $(DOCKER_BUILD_TOP)/proxy/envoy
+	time (cd $(DOCKER_BUILD_TOP)/proxy && \
 		docker build -t $(HUB)/proxy:$(TAG) -f Dockerfile.proxy .)
-endif
 
 docker.proxy_debug: tools/deb/envoy_bootstrap_tmpl.json
 docker.proxy_debug: ${ISTIO_ENVOY_DEBUG_PATH}
 docker.proxy_debug: $(ISTIO_OUT)/pilot-agent ${PROXY_JSON_FILES}
 docker.proxy_debug: pilot/docker/Dockerfile.proxy_debug
-	mkdir -p $(ISTIO_DOCKER)/proxyd
-	cp ${ISTIO_ENVOY_DEBUG_PATH} $(ISTIO_DOCKER)/proxyd/envoy
-	cp $^ $(ISTIO_DOCKER)/proxyd/
-	time (cd $(ISTIO_DOCKER)/proxyd && \
+	mkdir -p $(DOCKER_BUILD_TOP)/proxyd
+	cp ${ISTIO_ENVOY_DEBUG_PATH} $(DOCKER_BUILD_TOP)/proxyd/envoy
+	# Not using $^ to avoid 2 copies of envoy
+	cp tools/deb/envoy_bootstrap_tmpl.json $(ISTIO_OUT)/pilot-agent ${PROXY_JSON_FILES} pilot/docker/Dockerfile.proxy_debug $(DOCKER_BUILD_TOP)/proxyd/
```

Comment:
there is only 1 envoy in dependencies - revert?

Reply:
What do you mean ? There is one envoy in deps, and it is copied 2x, one by explicit line (cp ... .../envoy) and once by $^.
So we end up with 2 300M binaries, which need to be copied over to the docker env - 100s of times per day.

Reply:
sorry I must be missing something right in front of me
are you saying ${ISTIO_ENVOY_DEBUG_PATH} isn't a directory or marker and is different from pilot/docker/Dockerfile.proxy_debug ? why are there 2 ? (can't we use the one that has sha and `mv envoy-* envoy` in the staging dir ? why have 2 binaries ? (4 if counting non debug))

Reply:
I have no idea why we have to copy 2 300M binaries, and then copy 600M to the docker env (the full dir is copied). That's what I'm trying to fix. 

Sure, a mv would work as well, but so would changing  init.sh back to creating a file named 'release/envoy' ( and one in .../debug ). But someone seems to have a goal of turning make into 
bazel ( i.e. do everything and as obscure as it can be), and I can't seem to review (and block)
make related PRs fast enough.